[HttpKernel][FrameworkBundle] Add a minimalist default PSR-3 logger

dunglas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes <!-- don't forget updating src/**/CHANGELOG.md files -->
| BC breaks?    | no
| Deprecations? | no <!-- don't forget updating UPGRADE-*.md files -->
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/10321

This PR provides a minimalist PSR-3 logger that is always available when FrameworkBundle is installed.
By default, it writes errors on `stderr`, regular logs on `stdout` and discards debug data (this is configurable).

This approach has several benefits:

- It's what expect from an app logging systems of major containerization and orchestration tools including [Docker](https://docs.docker.com/engine/admin/logging/view_container_logs/) and [Kubernetes](https://kubernetes.io/docs/concepts/cluster-administration/logging/), as well as most cloud providers such as [Heroku](https://devcenter.heroku.com/articles/logging#writing-to-your-log) and [Google Container Engine](https://kubernetes.io/docs/tasks/debug-application-cluster/logging-stackdriver/). If the app follows this standard (and it's not currently the case with Symfony by default) logs will be automatically collected, aggregated and stored.
- It's in sync with the "back to Unix roots" philosophy of Flex
- Logs are directly displayed in the console when running the integrated PHP web server (`bin/console server:start` or Flex's `make serve`), Create React App also do that for instance.
- It fixes a common problem when installing Flex recipes: many bundles expect a logger service but currently there is none available by default, and you usually get a `"logger" service not found error` (because packages depend of the PSR, but the PSR doesn't provide a logger service).

dunglas committed:
[HttpKernel][FrameworkBundle] Add a minimalist default PSR-3 logger

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -303,6 +304,10 @@ public function load(array $configs, ContainerBuilder $container)
 
+        $loggerDefinition = $container->register(LoggerInterface::class, Logger::class);
+        $loggerDefinition->setPublic(false);
+        $container->setAlias('logger', LoggerInterface::class);
```
the alias also needs to be private.
for other services, we create the "noun" as a real definition, and the "fqcn" as the alias
for consistency, shouldn't we do the same here also?

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,110 @@
+final class Logger extends AbstractLogger
+{
+    /**
+     * @var array
```
should be removed

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,110 @@
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ *
+ * @see http://www.php-fig.org/psr/psr-3/
```
no need for that imho

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,110 @@
+    private $formatPattern;
+
+    /**
+     * @var bool
```
should be removed

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,110 @@
+
+    /**
+     * @param array  $levelToStream
+     * @param string $formatPattern
```
almost no value also, should be removed imho

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,110 @@
+     * @param array  $levelToStream
+     * @param string $formatPattern
+     */
+    public function __construct(array $levelToStream = array(), $formatPattern = '[%s] %s')
```
just "$format"? format+pattern looks redundant, isn't it?

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,110 @@
+     */
+    public function __construct(array $levelToStream = array(), $formatPattern = '[%s] %s')
+    {
+        $this->levelToStream = $levelToStream + self::$defaultLevelToStream;
```
destByLevel? outputByLevel?

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,110 @@
+            throw new InvalidArgumentException(sprintf('The log level "%s" does not exist.', $level));
+        }
+
+        if (!$this->errored) {
```
in fact, I'm skeptical about the need to write conditionally to stderr vs stdout.
despite its name, stderr is not only for errors but is for also for diagnostics, which all logs are to me,
especially in the simple logger we target.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,110 @@
+        }
+
+        if ($this->levelToStream[$level]) {
+            file_put_contents($this->levelToStream[$level], sprintf($this->formatPattern, $level, $this->interpolate($message, $context)).PHP_EOL, FILE_APPEND);
```
could it be "fwrite" for perf (ie the stream should be opened once?)

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,110 @@
+     *
+     * @return bool
+     */
+    public function hasErrored()
```
not sure we need the method?

dunglas replied:
In `ConsoleLogger`, it has been introduced by Nicolas in #19090. The use case is to be able to detect if an error has occurred and to set the appropriate exit code. But it doesn't look to be used in the Symfony code base and the PR doesn't give more information.
I've added this for the sake of consistency but maybe should it be removed?

nicolas-grekas replied:
OK, let's keep it :)

dunglas replied:
Ok, can you give me a bit more information about the use case? (this method isn't in any interface)

nicolas-grekas replied:
I dig a bit and we used to use it in the Blackfire Player - we don't anymore.
So we can consider it a non-common case and remove it from the simple logger we want here.
Let's drop :)

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,110 @@
+        }
+
+        // interpolate replacement values into the message and return
+        return strtr($message, $replace);
```
what about this for perf?
```php
        $replace = array();
        foreach ($context as $key => $val) {
            if (!\is_array($val) && (!\is_object($val) || method_exists($val, '__toString'))) {
                $replace["{{$key}}"] = $val;
            }
        }

        // interpolate replacement values into the message and return
        return $replace ? strtr($message, $replace) : $message;
```

ogizanagi replied:
Couldn't we reuse the same code as in [PsrLogMessageProcessor](https://github.com/Seldaek/monolog/blob/master/src/Monolog/Processor/PsrLogMessageProcessor.php) ?

dunglas replied:
I've merged both proposals and added some perf optims. I'll open a PR to update existing implementations like the one in `ConsoleLogger`.

nicolas-grekas reviewed (commented):
I like this idea :)

ogizanagi commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -303,6 +304,10 @@ public function load(array $configs, ContainerBuilder $container)
             $loader->load('web_link.xml');
         }
 
+        $loggerDefinition = $container->register(LoggerInterface::class, Logger::class);
```
Shouldn't it be in a pass and only if there is no `logger` service/alias already registered? (in other words, won't it conflict with the MonologBundle depending on the order the bundles are registered in the kernel? FrameworkBundle usually come first, but still)

fabpot replied:
The idea is indeed to provide this simple logger only when monolog (or any other logger) is not registered. Even if Flex tries to keep FrameworkBundle first, we should not rely on this.

ogizanagi reviewed (commented):
So, this PR aims to provide a sensible default logger, when not using MonologBundle, for instance when starting a simple project with flex.
I like it :)

I wonder: would it be interesting to allow not interpolating messages, but inline context at the end (see Monolog's LineFormatter)? It'll allow better aggregation, but I don't know if the logging systems you mentioned could leverage it easily.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,110 @@
+        LogLevel::ALERT => 'php://stderr',
+        LogLevel::CRITICAL => 'php://stderr',
+        LogLevel::ERROR => 'php://stderr',
+        LogLevel::WARNING => 'php://stdout',
```
Indeed, all logs should go to stderr IMHO.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,110 @@
+        LogLevel::ERROR => 'php://stderr',
+        LogLevel::WARNING => 'php://stdout',
+        LogLevel::NOTICE => 'php://stdout',
+        LogLevel::INFO => 'php://stdout',
```
I haven't had a look carefully, but I'm not sure we need to output info/notice logs. Intuitively, I would have skipped them.

dunglas commented:
I've changed the way levels are handled:

* Only `stderr` is used by default
* Levels less critical than `warning` are ignored by default
* if `kernel.debug` is set, all levels enabled (in the compiled pass, and sent to `stderr`)

I've also fixed most comments, thanks all for the review.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,119 @@
+        }
+
+        if (!$this->errored) {
+            $this->errored = in_array($level, array(LogLevel::EMERGENCY, LogLevel::ALERT, LogLevel::CRITICAL, LogLevel::ERROR));
```
can be \in_array + true as 3rd arg to trigger compile-time optims by the php enging

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/LoggerPass.php:
```diff
@@ -0,0 +1,45 @@
+     */
+    public function process(ContainerBuilder $container)
+    {
+        $loggerDefinition = $container->register('logger', Logger::class);
```
shouldn't this be set after an "if" somewhere? where is the "if necessary" condition?

ogizanagi replied:
It must check if there is no existing def/alias for `logger` and `LoggerInterface`.

nicolas-grekas reviewed (commented):
the logger should have a "reset" method, and be tagged "kernel.reset".

fabpot commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/LoggerPass.php:
```diff
@@ -0,0 +1,45 @@
+    {
+        $loggerDefinition = $container->register('logger', Logger::class);
+        $loggerDefinition->setPublic(false);
+        if ($container->getParameter('kernel.debug')) {
```
How does that work with the console. Wouldn't we have too much output now?

dunglas replied:
I'm not sure of what the best behavior is.

Imagine the following scenario:
* I've a problem in production
* I update the env var `APP_DEBUG` to `true`

I want all the available level of logs to be able to debug. (the current behavior).

However, Symfony is very verbose, `APP_DEBUG` is set to `true` for the `dev` env, and `dev` is the default env when the framework is installed. I agree that having a bunch of useless logs in the console when just running `make serve` is not a good developer experience.

What do you think about providing a custom logger definition (with less output) in the Flex recipe for just the `dev` env? We can also disable logs in the `test` env.

afurculita commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,119 @@
+                $replacements["{{$key}}"] = $val;
+            } elseif ($val instanceof \DateTimeInterface) {
+                $replacements["{{$key}}"] = $val->format($this->dateFormat);
+            } elseif (is_object($val)) {
```
Needs leading `\` too

afurculita commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,119 @@
+                $this->handles[$stream] = \fopen($stream, 'a');
+            }
+
+            \fwrite($this->handles[$stream], \sprintf($this->format, $level, $this->interpolate($message, $context)).PHP_EOL);
```
`PHP_EOL` needs a leading `\` too

nicolas-grekas replied:
we don't want to root-namespace everything, see https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3048 for some background.

stof replied:
Well, for constants, it would make sense though (as they are resolved at compile time for internal constants)

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,119 @@
+
+        if ($stream = $this->outputByLevel[$level]) {
+            if (!isset($this->handles[$stream])) {
+                $this->handles[$stream] = \fopen($stream, 'a');
```
`\` should be removed in fact, sorry for the confusion, only a short list of functions really benefit from `\`
see https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3048

dunglas commented:
@ogizanagi 

> I wonder: would it be interesting to allow not interpolating messages, but inline context at the end (see Monolog's LineFormatter)? It'll allow better aggregation, but I don't know if the logging systems you mentioned could leverage it easily.

AFAIK, most platforms doesn't support that unless you use their proprietary API (ex for Stackdriver: https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry).
I'm not aware of a standard way to do that (but if you have a hint, I'm very interested).

IMO we should keep this builtin logger as small and standard as possible and recommend to use other libraries such as Monolog for more advanced uses like this one... until a standard appears.

tyx commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,104 @@
+     *
+     * @return string
+     */
+    private function interpolate($message, array $context)
```
This PR is awesome ! But any chance to move this behavior in a dedicated class behind interface ? 
Like the Formatter of Monolog. I know you want to keep it simple but it is a big lack to be used imo.

We look for this kind of logger as it is our main usage (docker). But our logs are parsed into an ELK and without custom formatting we will not be able to use it (or change all other apps not in PHP that output logs and logstash config...).

dunglas replied:
Maybe can we just allow a callable in `format` instead of a string?

afurculita replied:
`interpolate` is added here only because it is part of the PSR-3 standard AFAIK. For custom advanced formatting, Monolog should be used

tyx replied:
I'm not sure a callable only for `format` would help (or I miss anything?). 

We would need both message and context separate as @ogizanagi said (I just realized he said the same thing).

I don't see any problem to not having this behavior embed by default, but if we could make it possible by design, we will definitively use it. Like a callable on the whole sprintf with `message` `level` `context` as arguments ?

But apparently other people are not enthusiastic about adding more configuration :p I guess we could just copy/paste the code also.

dunglas replied:
The callable will take the message and the context in parameter, and return a string. if none is passed, the current `sprintf` will be used.

It has several benefits:
- on one hand it's very flexible for the end user, allowing all kind of customizations
- on the other hand, we can remove all configuration options and it's very solid

WDYT @fabpot @javiereguiluz?

fabpot replied:
No configuration seems better to me. Any configuration should involve using Monolog.

ogizanagi replied:
Agreed. Let's forgot about this.

dunglas replied:
I've added a callable formatter implementation in a separate commit. It's just 1 line of code but it opens interesting possibility (see the json formatter in tests) in a cloud environment. IMO it is worth it.

shouze replied:
Without the formatter callable, I guess everybody will keep using monolog in fact, all the time. So in this case I'm not sure it's a good move to merge this PR at all.

Either monolog remains the log dependency - no need for this Logger - either the formatter should be added and we can sometimes decide to require monolog or not depending of the complexity of the wanted log stack. WDYT?

javiereguiluz commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,104 @@
+    private $dateFormat;
+    private $handles = array();
+
+    public function __construct(array $outputByLevel = array(), $format = '[%s] %s', $dateFormat = \DateTime::RFC3339)
```
Given that this is a minimalist logger, could we make some decisions about the formats instead of making them configurable? In other words, hardcode `[%s] %s` as `$format` and `\DateTime::RFC3339` as `$dateFormat`. If someone wants configurability, use a full-featured logger such as Monolog.

dunglas replied:
Why not, I've done that because it comes almost for free but we can be more restrictive. Exposing `$format` looks useful anyway.

fabpot replied:
I think I agree with @javiereguiluz. Even if it's free, it should be crystal clear that this logger is just a default/simple one. For any configurability, people still need to install Monolog. So, I would remove any way that allow configuration of this class.

fabpot commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,104 @@
+/**
+ * Minimalist PSR-3 logger designed to write in stdout, stderr or any other stream.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
```
Could be marked as @internal, right?

dunglas replied:
Indeed. It may be useful in other contexts than in FrameworkBundle, but this class is so simple that a copy/paste should be good enough.

stof commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,107 @@
+    private $handles = array();
+    private $formatter;
+
+    public function __construct(array $outputByLevel = array(), callable $formatter = null)
```
Why are you still allowing a formatter callable here, which is never used when wiring the service ?

dunglas replied:
https://github.com/symfony/symfony/pull/24300#discussion_r140799505

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:
```diff
@@ -22,6 +22,7 @@
 use Symfony\Bundle\FrameworkBundle\Command\YamlLintCommand;
 use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
 use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Bundle\FrameworkBundle\Logger\Logger;
```
what for?

jvasseur commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,107 @@
+        LogLevel::WARNING => 'php://stderr',
+        LogLevel::NOTICE => false,
+        LogLevel::INFO => false,
+        LogLevel::DEBUG => false,
```
Instead of using using by-level output streams I would defined an unique stream and a minimum logged level, I don't think logging different levels to separate streams is a common enough use case for a minimal logger.

dunglas replied:
I prefer to keep this feature, for instance it allows to redirect some levels on `stdout` and some others on `strerr`, or to create a file on disk per level.
This level to stream mapping and the formatter one allows to have a very flexible solution in just 100 LOC (100 LOC is what we can call minimal). For all more advanced use cases, Monolog must be installed and that's is.

fabpot replied:
redirecting logs to stdout is not a good practice and should not be encouraged, so that's not a valid use case IMHO. Creating a file is also not a use case. I tend to agree with @jvasseur, let's keep it minimal. It's not "Monolog should be installed for advanced use cases", it's more like "Monolog must be installed for any configurable logger". If not, drawing the line won't be easy.

dunglas replied:
I was looking for feature parity with the `log` package of the Go language (that is minimalist). This package has the ability to change the output (https://golang.org/pkg/log/#SetOutput) and to set the format. That's all.

nicolas-grekas replied:
Go only allows one single output, not one per level, isn't it?

dunglas replied:
Yes

shouze replied:
Makes sense to configure the output globally, not by level indeed.

dunglas commented:
Changed the implementation to:

* Only allow a minimal log level
* Only allow one input (as in Go)

jakzal commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,111 @@
+
+        $replacements = array();
+        foreach ($context as $key => $val) {
+            if (null === $val || is_scalar($val) || (\is_object($val) && method_exists($val, '__toString'))) {
```
Accessing function via the global namespace is inconsistent `\is_object` vs `is_scalar` etc. Any reasons for this?

dunglas replied:
Yes: https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3048 (`is_object` is in the list, `is_scalar` is not).

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+
+    public function __destruct()
```
we can reove this one, it will be done automatically by PHP

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,111 @@
+        }
+
+        $this->minLevelIndex = self::$levels[$minLevel];
+        $this->formatter = $formatter;
```
I propose to rename "interpolate()" to "format()", and replace this line by
`$this->formatter = $formatter ?: array($this, 'format');`

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,111 @@
+        $formatter = $this->formatter;
+        $message = $formatter ? $formatter($level, $message, $context) : sprintf('[%s] %s', $level, $this->interpolate($message, $context)).\PHP_EOL;
+
+        fwrite($this->handle, $message);
```
taking previous comment into account, this could be 
`fwrite($this->handle, $formatter($level, $message, $context));`

nicolas-grekas reviewed (approved):
with minor internal comments

jakzal commented on src/Symfony/Bundle/FrameworkBundle/Tests/Logger/LoggerTest.php:
```diff
@@ -0,0 +1,199 @@
+    {
+        $this->tmpFile = sys_get_temp_dir().'/log';
+        $this->logger = new Logger(LogLevel::DEBUG, $this->tmpFile);
+        file_put_contents($this->tmpFile, '');
```
We should at least try to clean up. In other tests saving files on the filesystem we use `tearDown` to remove them.

dunglas commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,99 @@
+            $message = strtr($message, $replacements);
+        }
+
+        return sprintf('[%s] %s', $level, $message).\PHP_EOL;
```
Should we add the date in the default formatter (like Go does)?

nicolas-grekas replied:
I guess we should. Or could the infra's log collector be responsible for adding the date? Is that common practice?

dunglas replied:
Docker doesn't do it by default at least.

shouze replied:
would be useful yes but default date format & precision should be decided in this case.

dunglas replied:
I would suggest RFC 3339 as this is a standard we use everywhere else.

shouze replied:
The only issue with `RFC 3339` is sometime you want a precision under the second but ok should be addressed by folks with the callable formatter for example.

nicolas-grekas replied:
RFC 3339 is the way to go to me also, and a custom formatter/monolog would overdo it for anyone needing more/another format.

Taluu commented on src/Symfony/Bundle/FrameworkBundle/Logger/Logger.php:
```diff
@@ -0,0 +1,99 @@
+
+        $this->minLevelIndex = self::$levels[$minLevel];
+        $this->formatter = $formatter ?: array($this, 'format');
+        if (false === $this->handle = @fopen($output, 'a')) {
```
why leave open a handle in the constructor ? Why not use `file_put_contents` which opens / closes the handle each time it is needed ?

dunglas replied:
For performance.

Taluu replied:
But for something that is used as a daemon, isn't this more costly to leave an handle open for a long while, rather than opening / closing it ?

A middleground could be to open the handle only on the first log call (if no logs, no need to open it) ?

dunglas replied:
It was my first implementation, but after a discussion with @nicolas-grekas, we agreed that it costs almost nothing to open a handle for stderr.

Taluu replied:
Ok, fair enough

nicolas-grekas replied:
opening in the constructor gives the possibility to throw an exception in the constructor, rather that "who knows where - an nowhere fits, this is a logger".
opening lazilly is not worth it IMHO, that's a single fopen.
closing for long running process would not make them faster, just slower in fact (but that'd be negligible yes.)

fabpot commented:
I'm wondering if the Logger shouldn't be part of the HttpKernel component instead. We had a logger there in the past, and as there is nothing tied to FrameworkBundle, it might be better (for reusability). WDTY? We could even move it into a new component, but that might be too much.

shouze commented:
@fabpot yes in fact `HttpKernel` component should probably be split into `Kernel` and `HttpKernel` at some point because when you want to create a cli, this `Http` in `HttpKernel` should not be a requirement. It's another debate of course and it's a very good idea to put the minimal logger in the logger BTW.

dunglas commented:
@fabpot excellent idea. It will allow to easily reuse this tiny logger without having to install the whole framework.

stof commented:
I agree about moving it to the HttpKernel component.

I don't think a dedicated component makes sense, as we don't plan to turn this into a configurable logger (Monolog is there for that), and the interface is not part of it either (psr/log is there)

fabpot commented:
Ok, let's move it to HttpKernel then.

dunglas commented:
- Moved to HttpKernel
- Marked as `@experimental` instead of `@internal` (I think it makes sense with the move, but let me know if you prefer I revert that)
- Changelogs updated

fabpot commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/LoggerPass.php:
```diff
@@ -0,0 +1,45 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler;
```
Would it make sense to move it to HttpKernel as well?

dunglas replied:
done

fabpot commented on src/Symfony/Component/HttpKernel/Log/Logger.php:
```diff
@@ -0,0 +1,99 @@
+/**
+ * Minimalist PSR-3 logger designed to write in stderr or any other stream.
+ *
+ * @experimental
```
We are only using `@experimental` when we are very unsure about a feature. That's not the case here. So `@internal` seems better to me. Or nothing at all :)

dunglas replied:
Ok, I think `@internal` is better for now. We can always  remove this annotation in the future.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/FrameworkBundle.php:
```diff
@@ -119,6 +120,7 @@ public function build(ContainerBuilder $container)
         $this->addCompilerPassIfExists($container, FormPass::class);
         $container->addCompilerPass(new WorkflowGuardListenerPass());
         $container->addCompilerPass(new ResettableServicePass());
+        $container->addCompilerPass(new LoggerPass());
```
should it be registered with a lower priority?

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Log/Logger.php:
```diff
@@ -0,0 +1,99 @@
+/**
+ * Minimalist PSR-3 logger designed to write in stderr or any other stream.
+ *
+ * @experimental
```
I would remove this and not put `@internal` either, this should be just a regular class to me, that'd be fine eg decorating in userland if ones want to.

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Log/Logger.php:
```diff
@@ -0,0 +1,99 @@
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+final class Logger extends AbstractLogger
```
no need to make it final: all methods are in the interface so inheritance is really not an issue for maintenance.

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Log/Logger.php:
```diff
@@ -0,0 +1,99 @@
+
+    /**
+     * @param string $level
+     * @param string $message
```
missing $context

dunglas replied:
On purpose, the type is already documented by the typehint.

nicolas-grekas replied:
when I did so on some of my PRs, stof told me we don't do partial doc blocks :)

nicolas-grekas replied:
but we ma also remove the docblock entirely :)

dunglas replied:
Other types aren't inferable.

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Log/Logger.php:
```diff
@@ -0,0 +1,99 @@
+/**
+ * Minimalist PSR-3 logger designed to write in stderr or any other stream.
+ *
+ * @internal
```
should be removed

dunglas replied:
Sure? Why not but https://github.com/symfony/symfony/pull/24300#discussion_r141750369

nicolas-grekas replied:
Yep, I don't agree this should be internal.
We're writing a framework, we should provide open/closed building blocks, not only closed ones.
Here, there is no reason to make the class neither internal not final: all methods are in a PSR interface.

dunglas replied:
The problem here is the constructor. We're still not very sure about it. But removed, it should be easy to deprecate some arguments if needed.

fabpot replied:
Marking it `@internal` is important IMHO. Not for technical reasons, just to reinforce the idea that this is a simple logger that should not be extended/decorated. If the need is more complex, people should move to Monolog instead.

fabpot replied:
my comment was more about `final`, sorry :)

dunglas replied:
So `final`, `@internal` or both? I vote for both.

nicolas-grekas replied:
Internal means we can rename the class as we want, that's not the case to me.
If I want to create a lazy version of this logger, I should be allowed to.
So *at most*, `@final` to me even if I see no good reason it should, as explained.

dunglas replied:
Hmm, after a second though `@internal` doesn't make sens now that we can inject a formatter.

fabpot replied:
Ok, let's remove both @final and @internal.

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Log/Logger.php:
```diff
@@ -0,0 +1,99 @@
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+final class Logger extends AbstractLogger
```
"final" should be removed to me, it doesn't provide anything in terms of design

dunglas replied:
I usually prefer to be defensive by default but indeed here all methods are part of the interface.

dunglas committed:
Remove final and @internal. Change pass priority.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/FrameworkBundle.php:
```diff
@@ -85,6 +86,7 @@ public function build(ContainerBuilder $container)
     {
         parent::build($container);
 
+        $container->addCompilerPass(new LoggerPass());
```
sorry, I meant something like:
`new LoggerPass(), PassConfig::TYPE_BEFORE_OPTIMIZATION, -32`
but maybe order doesn't/shouldn't matter in fact?

dunglas replied:
I think it doesn't matter, but I'm not 100% sure. Will add your patch.

dunglas committed:
Add @param. Change pass priority

nicolas-grekas reviewed (approved):
:+1: as is :)

nicolas-grekas commented:
when test are reasonably green :)

dunglas committed:
Fix tests

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/FrameworkBundle.php:
```diff
@@ -18,6 +18,7 @@
 use Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler\CachePoolClearerPass;
 use Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler\CachePoolPrunerPass;
 use Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler\DataCollectorTranslatorPass;
+use Symfony\Component\HttpKernel\DependencyInjection\LoggerPass;
```
alpha order

nicolas-grekas commented on src/Symfony/Component/HttpKernel/Tests/Log/LoggerTest.php:
```diff
@@ -0,0 +1,212 @@
+        $this->logger->log(LogLevel::DEBUG, 'test', array('user' => 'Bob'));
+
+        // Will always be true, but asserts than an exception isn't thrown
+        $this->assertLogsMatch(array(), $this->getLogs());
```
should be `assertSame(array(), ...`

fabpot commented:
Thank you @dunglas
